### ISSUE TITLE
Builds exceeding lock not scheduled

### ISSUE THREAD

**Issue by vbraun** [buildbot/buildbot#4491]
We have a lock limiting buildbot to 3 concurrent build jobs. Submitted tasks exceeding the limit used to be shown with a yellow badge on /#/builders (just like builds in progress), but now not any more. This may be an intentional change, though it doesn't really make sense to me. 

The pending tasks are shown on /#/pendingbuildrequests. The pending tasks are not being started when the lock is being released by a completed task, so the pending tasks never start. This is certainly a bug. 

This is the sagemath buildbot http://build.sagemath.org running buildbot-1.7.0 on python 2.7

There are two exceptions it twistd.log (attached), though they don't seem to be obviously related:
[bulidbot-twistd.log](https://github.com/buildbot/buildbot/files/2708967/bulidbot-twistd.log)

**Comment by vbraun** [comment]
For the record, the the debug tab on #/buildrequests/<n> for such a pending-but-never-starting request says
~~~~
Buildrequest
builderid 71 
buildrequestid 23 
buildsetid 23 
claimed false 
claimed_at
claimed_by_masterid
complete false 
complete_at
priority 0 
properties
results -1 
submitted_at 1545760877 
waited_for false 

Buildset 
bsid 23 
complete false 
complete_at
external_idstring
parent_buildid
parent_relationship
reason A build was forced by 'sage': force build 
results -1 
sourcestamps [{"branch":"","codebase":"","created_at":1545735534,"patch":null,"project":"","repository":"","revision":"","ssid":1}]
submitted_at 1545760877 

Builder
_endpoint builders 
builderid 71 
description
masterids [1] 
name Zpkg slow AIMS bu18_64 (Ubuntu 18.04 64 bit) pkg-binary 
tag ["binary-pkg"] 
_id 71 

Builds
~~~~

**Comment by vbraun** [comment]
Downgrading to buildbot-1.6.0 fixes it for me

**Comment by p12tic** [comment]
Likely caused by my changes. Will look ASAP.

**Comment by p12tic** [comment]
Hey @vbraun, could you share the buildbot config that you're using? So far I've been unable to reproduce the issue of builds being not scheduled on a simple setup (3 workers, 1 builder and a master lock limiting build concurrency to 2). Once the lock becomes available, subsequent builds are scheduled as expected.

On Buildbot 1.6 and before, we scheduled builds as soon as builder became available without any regard to lock availability. This approach caused issues, because in the case of latent workers it's not guaranteed that a worker can build a specific build at specific time. So on 1.7 and onwards, a build is scheduled only once it can actually run: i.e. all locks can be acquired and there's a free and compatible worker. If a build can't satisfy these conditions, it will live in the list of pending build requests.

If you're indeed seeing build requests that are never started even after the lock becomes free, it's indeed a bug.

**Comment by vbraun** [comment]
The lock configuration is just 
{{{
from buildbot.locks import MasterLock
AIMS_VM_LOCK = MasterLock("AIMS Virtual Machine Lock", maxCount=3)
}}}
I'll PM you the entire buildbot config

**Comment by p12tic** [comment]
With the help of @vbraun I've been able to reproduce the issue locally and later fix it. Thank you!

**Comment by p12tic** [comment]
Fixed in #4525.
